### PR TITLE
VSR: Fix replicas=2 abdiction deadlock

### PR DESCRIPTION
**Pull request by sentientwaffle**
## Pre-merge checklist

Performance:

* [x] I am very sure this PR could not affect performance.

### CODE REVIEW

File: src/vsr/replica.zig [hunk 1]
```diff
@@ -1847,6 +1847,18 @@ pub fn ReplicaType(
             self.primary_abdicate_timeout.reset();
             if (self.replica_count == 1) return;
 
+            // When replica_count=2 the primary never abdicates to avoid the following deadlock:
+            // 1. Primary is in status=normal for view V.
+            // 2. Backup is in status=view_change for view V.
+            // 3. Backup missed the primary's start_view.
+            // 4. Primary is trying to prepare a message from V-1.
```
Comment (matklad) [inline src/vsr/replica.zig]: Hm, why primary prepares stuff from V-1 if it's status=normal in V?

Comment (sentientwaffle) [reply src/vsr/replica.zig]: The message from V-1 is "maybe-committed, maybe-uncommitted", so the new primary needs to treat it as if it is in fact committed.

**Comment by sentientwaffle** [comment]
Converting this to draft. I'm going to take a different approach which avoids the replica-count special-case.

**Comment by sentientwaffle** [comment]
Rewritten. Now a backup in view-change status periodically sends a `request_start_view`.

One alternate approach I considered:

Since during that state we are _also_ sending a DVC (for the same view we are requesting), maybe the primary could treat a DVC for an already-started view as if it _was_ a request-start-view.

I chose the timeout approach b/c it lends itself better to message debouncing, but don't feel strongly about it.

**Comment by matklad** [comment]
Like the overall approach, not completely sold on details though.

>Since during that state we are also sending a DVC (for the same view we are requesting), maybe the primary could treat a DVC for an already-started view as if it was a request-start-view.

Replying to DVC with SV does seem a bit more elegant. The drawback is that this sends some needless SVs (for replicas outside DVC quorum, we'll send SVs at lest twice). 

With debountincing, I guess the ideal code would look like this:

```
fn on_DVC_or_RSV_timeout():
    if (has seen new view):
        send request_start_view
    else:
        send do_view_change
        send request_start_view
```

That is, it makes sense to re-use the same timer here, because we know that, if we should send RSV, sending DVC for older view would be useless.


I guess probably makes sense to land as is, and reconsider once we put debouncing in

File: src/vsr/replica.zig [hunk 2]
```diff
@@ -5533,6 +5564,7 @@ pub fn ReplicaType(
                 self.start_view_change_message_timeout.start();
                 self.view_change_status_timeout.stop();
                 self.do_view_change_message_timeout.stop();
+                self.request_start_view_message_timeout.stop();
```
Comment (matklad) [inline src/vsr/replica.zig:5568]: I think we can assert that that timeout was ticking?

Other than this, r=me

**Pull request merged**